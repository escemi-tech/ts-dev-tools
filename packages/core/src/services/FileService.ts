import { copyFileSync, existsSync, readFileSync, writeFileSync } from "fs";

export class FileService {
  static fileExists(filePath: string): boolean {
    return existsSync(filePath);
  }

  static getFileContent(filePath: string): string {
    return readFileSync(filePath, "utf-8");
  }

  static putFileContent(path: string, content: string) {
    writeFileSync(path, content);
  }

  static copyFile(sourcePath: string, destinationPath: string) {
    copyFileSync(sourcePath, destinationPath);
  }
}
