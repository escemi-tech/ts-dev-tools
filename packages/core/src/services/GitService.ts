import { chmodSync, createWriteStream, existsSync, statSync } from "fs";
import { join } from "path";

import { PROJECT_NAME, PROJECT_URL } from "../constants";
import { CmdService } from "./CmdService";

export class GitService {
  static GIT_HOOK_MODE = 0o755;
  static GIT_HOOK_TEMPLATE = `#!/bin/sh

# Created by ${PROJECT_NAME} (${PROJECT_URL})

%gitHookCommand%`;

  static async isGitRepository(absoluteProjectDir: string): Promise<boolean> {
    return await CmdService.execCmd("git rev-parse", absoluteProjectDir, true)
      .then(() => true)
      .catch(() => false);
  }

  static async addGitHook(
    absoluteProjectDir: string,
    gitHookName: string,
    gitHookCommand: string
  ): Promise<void> {
    const gitHookDirPath = join(absoluteProjectDir, ".git/hooks");
    const gitHookFilePath = join(gitHookDirPath, gitHookName);

    if (existsSync(gitHookFilePath)) {
      const mode = GitService.getFilePermissions(gitHookFilePath);
      if (mode !== GitService.GIT_HOOK_MODE) {
        chmodSync(gitHookFilePath, GitService.GIT_HOOK_MODE);
      }
      return;
    }

    return new Promise((resolve, reject) => {
      const stream = createWriteStream(gitHookFilePath, { mode: GitService.GIT_HOOK_MODE });
      stream.on("close", () => resolve(undefined));
      stream.on("error", (error) => reject(error));
      stream.write(GitService.GIT_HOOK_TEMPLATE.replace("%gitHookCommand%", gitHookCommand));
      stream.end();
    });
  }

  private static getFilePermissions(filePath: string) {
    const { mode } = statSync(filePath);
    return mode & parseInt("777", 8);
  }
}
