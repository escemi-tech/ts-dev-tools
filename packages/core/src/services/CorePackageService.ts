import { resolve } from "path";

import { PackageJson } from "./PackageJson";

export class CorePackageService {
  static getPackageRootPath() {
    return resolve(__dirname, "../..");
  }

  static getPackageName() {
    return PackageJson.fromDirPath(CorePackageService.getPackageRootPath()).getPackageName();
  }
}
