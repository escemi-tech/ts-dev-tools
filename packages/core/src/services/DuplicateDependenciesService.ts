import { PackageJson } from "./PackageJson";
import { Plugin, PluginService } from "./PluginService";

type DuplicateDependencies = Map<string, Set<string>>;

export class DuplicateDependenciesService {
  static duplicateDependencies(absoluteProjectDir: string): void {
    console.info(`Checking for duplicate dev dependencies...`);

    const duplicateDependencies: DuplicateDependencies = new Map();
    DuplicateDependenciesService.getProjectDuplicateDependencies(
      absoluteProjectDir,
      duplicateDependencies
    );

    DuplicateDependenciesService.printDuplicateDependencies(duplicateDependencies);

    console.info(`Check for duplicate dev dependencies done!`);
  }

  private static getProjectDuplicateDependencies(
    absoluteProjectDir: string,
    duplicateDependencies: DuplicateDependencies
  ) {
    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    const projectPackageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const projectDevDependencies = projectPackageJson.getDevDependenciesPackageNames();

    for (const plugin of installedPlugins) {
      DuplicateDependenciesService.getPluginDuplicateDependencies(
        plugin,
        projectDevDependencies,
        duplicateDependencies
      );
    }

    return duplicateDependencies;
  }

  private static getPluginDuplicateDependencies(
    plugin: Plugin,
    projectDevDependencies: string[],
    duplicateDependencies: DuplicateDependencies
  ) {
    // First check for duplicate of inherited plugins
    DuplicateDependenciesService.getProjectDuplicateDependencies(
      plugin.path,
      duplicateDependencies
    );

    const pluginPackageJson = PackageJson.fromDirPath(plugin.path);
    const pluginDependencies = pluginPackageJson.getDependenciesPackageNames();

    let pluginDuplicateDependencies = duplicateDependencies.get(plugin.fullname);
    if (!pluginDuplicateDependencies) {
      pluginDuplicateDependencies = new Set();
      duplicateDependencies.set(plugin.fullname, pluginDuplicateDependencies);
    }

    for (const projectDevDependency of projectDevDependencies) {
      if (pluginDependencies.includes(projectDevDependency)) {
        pluginDuplicateDependencies.add(projectDevDependency);
      }
    }
  }

  private static printDuplicateDependencies(duplicateDependencies: DuplicateDependencies) {
    duplicateDependencies.forEach((pluginDuplicateDependencies, plugin) => {
      if (pluginDuplicateDependencies.size) {
        const pluginDuplicateDependenciesValue = Array.from(pluginDuplicateDependencies);
        console.info(
          `Some dev dependencies are unnecessarily installed as their are already required by "${plugin}":"\n  - "${pluginDuplicateDependenciesValue.join(
            "\n  - "
          )}\n`
        );
      }
    });
  }
}
