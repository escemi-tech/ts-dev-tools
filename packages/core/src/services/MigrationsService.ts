import { readdirSync, unlinkSync } from "fs";
import { resolve } from "path";

import { PackageJson } from "../services/PackageJson";
import { Plugin, PluginService } from "./PluginService";

export type Migration = { fullname: string; shortname: string; path: string };

export type MigrationUpFunction = (absoluteProjectDir: string) => Promise<void>;

export class MigrationsService {
  static MIGRATION_BUILT_PATH = "dist/install/migrations";

  static async executeMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Promise<void> {
    const migrations = MigrationsService.getAvailableMigrations(absoluteProjectDir, currentVersion);

    const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const packageJsonBackupPath = packageJson.backup();

    try {
      for (const migration of migrations) {
        console.info(`Applying migration "${migration.fullname}"...`);

        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { up } = require(migration.path);

        // Apply migration
        await up(absoluteProjectDir);

        // Upgrade current version
        PackageJson.fromDirPath(absoluteProjectDir).merge({
          tsDevTools: { version: migration.shortname },
        });

        console.info(`Migration "${migration.fullname}" applied!`);
      }
    } catch (error) {
      // Rollback package.json
      packageJson.restore(packageJsonBackupPath);
      throw error;
    }

    unlinkSync(packageJsonBackupPath);
  }

  private static getAvailableMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Migration[] {
    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    const migrationFiles: Migration[] = [];
    for (const installedPlugin of installedPlugins) {
      migrationFiles.push(
        ...MigrationsService.getPluginMigrations(installedPlugin, currentVersion)
      );
    }

    migrationFiles.sort(({ shortname: nameA }, { shortname: nameB }) => nameA.localeCompare(nameB));

    return Array.from(new Set(migrationFiles));
  }

  private static getPluginMigrations(
    plugin: Plugin,
    currentVersion: string | undefined
  ): Migration[] {
    const migrationFiles = [];

    const pluginMigrationsDirPath = resolve(plugin.path, MigrationsService.MIGRATION_BUILT_PATH);
    for (const migrationFile of readdirSync(pluginMigrationsDirPath)) {
      if (!migrationFile.match(/^[0-9]{14}-[-a-z]+\.(js|ts)$/)) {
        continue;
      }

      const migrationName = MigrationsService.getMigrationNameFromFile(migrationFile);
      const shouldApplyMigration = MigrationsService.migrationIsAfterCurrentVersion(
        migrationName,
        currentVersion
      );

      if (!shouldApplyMigration) {
        continue;
      }

      migrationFiles.push({
        shortname: migrationName,
        fullname: `${plugin.shortname} - ${migrationName}`,
        path: resolve(pluginMigrationsDirPath, migrationFile),
      });
    }
    return migrationFiles;
  }

  public static getMigrationNameFromFile(migrationFile: string): string {
    const migrationName = migrationFile.split(".").slice(0, -1).join(".");
    return migrationName;
  }

  public static migrationIsAfterCurrentVersion(migrationName: string, currentVersion?: string) {
    return !currentVersion || currentVersion.localeCompare(migrationName) < 0;
  }
}
