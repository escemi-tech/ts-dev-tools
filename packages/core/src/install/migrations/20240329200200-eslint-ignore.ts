import { MigrationUpFunction } from "../../services/MigrationsService";
import { PackageJson } from "../../services/PackageJson";

export const up: MigrationUpFunction = async (absoluteProjectDir: string): Promise<void> => {
  const eslintConfig = {
    ignorePatterns: ["dist", "node_modules"],
  };

  const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
  packageJson.merge({
    eslintConfig,
  });
};
