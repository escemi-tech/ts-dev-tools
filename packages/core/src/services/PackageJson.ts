import { PackageJson as PackageJsonType } from "type-fest";

import { PackageJsonMerge } from "./PackageJsonMerge";
import { FileService } from "./FileService";
import { join } from "path";

export type JsonArray = boolean[] | number[] | string[] | JsonFileData[] | Date[];
export type AnyJson = boolean | number | string | JsonFileData | Date | JsonArray | JsonArray[];

export type JsonFileData = {
  [key: string]: AnyJson | undefined;
};
export type PackageJsonContent = JsonFileData & PackageJsonType;

export class PackageJson {
  private static readonly PACKAGE_JSON_FILE_NAME = "package.json";

  private content?: PackageJsonContent = undefined;

  constructor(private path: string) {
    if (!FileService.fileExists(this.path)) {
      throw new Error(`Package.json "${this.path}" does not exist`);
    }
  }

  getPath(): string {
    return this.path;
  }

  getContent(): PackageJsonContent {
    if (this.content) {
      return this.content;
    }
    return (this.content = JSON.parse(FileService.getFileContent(this.path)) as PackageJsonContent);
  }

  setContent(content: PackageJsonContent): void {
    this.content = content;
    this.write();
  }

  getPackageName(): string | undefined {
    return this.getContent().name;
  }

  getPackageVersion(): string | undefined {
    return this.getContent().version;
  }

  isPrivate(): boolean {
    return this.getContent().private === true;
  }

  getTsDevToolsVersion(): string | undefined {
    const tsDevToolsConfig = this.getContent().tsDevTools as JsonFileData | undefined;
    const version = tsDevToolsConfig?.version as string | undefined;
    return version;
  }

  getDependenciesPackageNames(): string[] {
    const dependencies = this.getContent().dependencies;
    return dependencies ? Object.keys(dependencies) : [];
  }

  getDevDependenciesPackageNames(): string[] {
    const devDependencies = this.getContent().devDependencies;
    return devDependencies ? Object.keys(devDependencies) : [];
  }

  getAllDependenciesPackageNames(): string[] {
    return Array.from(
      new Set([...this.getDependenciesPackageNames(), ...this.getDevDependenciesPackageNames()])
    );
  }

  merge(update: PackageJsonContent): void {
    this.content = PackageJsonMerge.merge(this.getContent(), update);
    this.write();
  }

  backup(): string {
    const backupPath = this.path + ".backup";
    FileService.copyFile(this.path, backupPath);
    return backupPath;
  }

  restore(backupPath: string): void {
    FileService.copyFile(backupPath, this.path);
    this.content = undefined;
  }

  private write() {
    FileService.putFileContent(this.path, JSON.stringify(this.content, null, 2));
  }

  static fromDirPath(dirPath: string): PackageJson {
    const packageJsonPath = join(dirPath, PackageJson.PACKAGE_JSON_FILE_NAME);
    return new PackageJson(packageJsonPath);
  }
}
