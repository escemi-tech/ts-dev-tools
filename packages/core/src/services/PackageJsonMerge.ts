import { AnyJson, JsonArray, JsonFileData, PackageJsonContent } from "./PackageJson";

export class PackageJsonMerge {
  static merge(source: PackageJsonContent, update: PackageJsonContent): PackageJsonContent {
    return PackageJsonMerge.mergeObjects(source, update);
  }

  private static mergeValues(
    source: AnyJson | undefined,
    update: AnyJson | undefined
  ): AnyJson | undefined {
    if (source === undefined) {
      return update;
    }

    if (!PackageJsonMerge.typesEqual(source, update)) {
      throw new Error(`Unable to merge package json value because types are different`);
    }

    // Deal with arrays
    if (Array.isArray(update)) {
      return PackageJsonMerge.mergeArrays(source as JsonArray[], update as JsonArray[]);
    }

    // Deal with objects
    if (typeof update === "object") {
      return PackageJsonMerge.mergeObjects(
        source as PackageJsonContent,
        update as PackageJsonContent
      );
    }

    return update;
  }

  private static mergeObjects(source: JsonFileData, update: JsonFileData): JsonFileData {
    for (const updateKey in update) {
      if (!Object.prototype.hasOwnProperty.call(update, updateKey)) {
        continue;
      }
      source[updateKey] = PackageJsonMerge.mergeValues(source[updateKey], update[updateKey]);
    }
    return source;
  }

  private static mergeArrays(source: JsonArray[], update: JsonArray[]): JsonArray[] {
    for (const item of update) {
      if (!source.includes(item)) {
        source.push(item);
      }
    }
    return source;
  }

  private static typesEqual(sourceValue: unknown, updateValue: unknown) {
    if (Array.isArray(sourceValue) && !Array.isArray(updateValue)) {
      return false;
    }

    return typeof sourceValue === typeof updateValue;
  }
}
