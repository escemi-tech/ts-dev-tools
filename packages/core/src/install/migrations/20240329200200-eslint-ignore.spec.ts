import { PackageJson } from "../../services/PackageJson";
import {
  createTestProjectDirWithFixtures,
  removeTestProjectDir,
  restorePackageJson,
} from "../../tests/project";
import { up } from "./20240329200200-eslint-ignore";

describe("Migration 20240329200200-eslint-ignore", () => {
  let testProjectDir: string;

  beforeAll(() => {
    testProjectDir = createTestProjectDirWithFixtures(__filename);
  });

  afterAll(() => {
    removeTestProjectDir(__filename);
  });

  describe("Up", () => {
    afterEach(() => {
      restorePackageJson(__filename);
    });

    it("should apply migration", async () => {
      await up(testProjectDir);

      const packageJsonContent = PackageJson.fromDirPath(testProjectDir).getContent();

      expect(packageJsonContent).toMatchSnapshot();
    });
  });
});
