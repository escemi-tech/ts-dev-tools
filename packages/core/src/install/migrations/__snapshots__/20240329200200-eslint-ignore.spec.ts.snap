// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migration 20240329200200-eslint-ignore Up should apply migration 1`] = `
{
  "eslintConfig": {
    "ignorePatterns": [
      "dist",
      "node_modules",
    ],
  },
  "version": "1.0.0",
}
`;
